CIP-????/README.md

```diff
@@ -0,0 +1,150 @@
+> write validation logic to set constraints for the creation and transfer of Value and Data across
+> the Cardano blockchain. **Given that half of the entire purpose of this language is to express
+> constraints over the creation and transfer of** `Value`, why is Value treated as a standard
+> library concept rather than a first-class language primitive?
+
+This was the motivation for [CIP-0153][1]. Given the overwhelming importance for validating `Value`,
+it makes sense for `BuiltinValue` to have a reasonably disproportionate number of builtin operations
+as compared to `BuiltinList` or `BuiltinArray`. The current builtins added in CIP-0153 do not cover
+many of the use cases for `Value`. As a reference, see the Aiken stdlib's [functions on `Value`][2].
+Functions like `tokens`, `negate`, and `policies` are not possible to implement using the CIP-0153
+builtin functions which means they will not benefit from the new `BuiltinValue` type. Plutus needs
+enough builtins to cover the main operations on `BuiltinValue`, either through generalizable
+builtins or dedicated ones.
+
+## Specification
+
+This CIP proposes adding the following new builtins functions:
+
+```haskell
+type BuiltinCurrencySymbol = BuiltinByteString
+
+-- | Negate all values in a `BuiltinValue`.
+negate :: BuiltinValue -> BuiltinValue
+
+-- | Intersection of two `BuiltinValue`s. Returns data in the first value for the (policy ids, token
+-- names) existing in both values.
+intersection :: BuiltinValue -> BuiltinValue -> BuiltinValue
+
+-- | Returns all policy ids found in the value.
+policies :: BuiltinValue -> List [BuiltinCurrencySymbol]
+
+-- | Return all tokens and their quantities for a given policy id. It returns a `BuiltinValue` so
+-- that the result can make use of the `lookupCoin` builtin added in CIP-0153. It can always be
+-- converted to a `List` for pattern-matching using the `valueData` builtin added in CIP-0153.
+-- See the Rationale section for why `intersection` isn't used instead.
+lookupTokens :: BuiltinCurrencySymbol -> BuiltinValue -> BuiltinValue
+```
+
+## Rationale: how does this CIP achieve its goals?
+
+After this CIP, the total set of `BuiltinValue` functions will be:
+
+```haskell
+-- CIP-0153 functions
+insertCoin :: BuiltinCurrencySymbol -> BuiltinTokenName -> BuiltinInteger -> BuiltinValue -> BuiltinValue
+lookupCoin :: BuiltinValue -> BuiltinCurrencySymbol -> BuiltinTokenName -> BuiltinInteger
+valueContains :: BuiltinValue -> BuiltinValue -> Bool
+unionValue :: BuiltinValue -> BuiltinValue -> BuiltinValue
+valueData :: BuiltinValue -> BuiltinData
+unValueData :: BuiltinData -> BuiltinValue
+
+-- New functions
+negate :: BuiltinValue -> BuiltinValue
+intersection :: BuiltinValue -> BuiltinValue -> BuiltinValue
+policies :: BuiltinValue -> List [BuiltinCurrencySymbol]
+lookupTokens :: BuiltinCurrencySymbol -> BuiltinValue -> BuiltinValue
+```
+
+From these builtin functions, most of Aiken's stdlib `Value` functions can now make use of the
```

Comment:
To be a bit annoying: it would make this proposal much more concrete to have an appendix listing all the interesting functions you want to implement with their implementations using the new primitives. I think they're presumably all quite simple, and it would make the case quite compelling.

_Ideally_ we'd also be able to see how this improves the costing behaviour of these functions.